### ISSUE TITLE
[#85 M7] Eval harness, cost/latency metrics, and Shadow→Verify→Skip rollout

### ISSUE THREAD

**Issue by Exios66** [LLM-Mailroom-Services/mailroom-issues#92]
## Parent
Child of [#85](https://github.com/LLM-Mailroom-Services/mailroom-issues/issues/85).

**Related:** [#84](https://github.com/LLM-Mailroom-Services/mailroom-issues/issues/84) thresholds + skip-sorter end-state; promote only after eval sign-off.

## Goal
Phase 4: eval harness + cost/latency/disagreement metrics, then ordered rollout Shadow → Verify → Skip (allowlisted classes).

### Eval & metrics
- Short-doc cohort eval (mailroom-dataset / held-out slice): agreement vs sorter, calibrated ECE, $/doc, p50/p95 latency
- Dashboards/panels: BERT-pass rate, sorter-skip rate, disagreement rate, fail-soft rate, downstream extract accuracy
- Langfuse/Phoenix + ops hooks for BERT path health

### Rollout
1. **Shadow** — BERT always computed; sorter always runs; metrics only
2. **Verify** — gate on; PASS requires agreement; no skip
3. **Skip** — PASS skips LLM sorter for allowlisted classes only (start correspondence / short notices)
- Document cost reduction on short-doc cohort without accuracy regression beyond agreed bound
- Rollback: `MAILROOM_BERT_INTAKE=0`; mode flag demotion

## Out of scope
- Training new checkpoints (#84)
- Taxonomy catalog redesign

## Acceptance criteria / DoD
- [ ] Eval suite runnable in CI or documented dojo path with published metrics schema
- [ ] Shadow eval complete on short-doc cohort
- [ ] Verify canary green (disagreement → LLM authority)
- [ ] Skip enabled only for allowlisted classes with cost/accuracy sign-off recorded on this issue
- [ ] Agents README / intake doctrine + rollback documented
- [ ] Dashboard panels live (or tracked follow-up with owners named)

## Depends on
- [#88](https://github.com/LLM-Mailroom-Services/mailroom-issues/issues/88), [#90](https://github.com/LLM-Mailroom-Services/mailroom-issues/issues/90), [#91](https://github.com/LLM-Mailroom-Services/mailroom-issues/issues/91)
- [#84](https://github.com/LLM-Mailroom-Services/mailroom-issues/issues/84) calibration for production skip

## Blocks
- Nothing further under #85 (epic completion when this closes)

**Comment by Exios66** [comment]
**Progress (2026-09-18):** eval harness landed — `training/eval_modernbert.py` (plan §11 surfaces 1-4): held-out test split only, stratified sampling, doc_type/subclass accuracy, per-stratum confusion, window ECE + band ECE, per-head temperatures, and `--selective-risk` threshold sweep (plan §8). Tracing spans in `src/mailroom_ml/tracing.py`. NOT yet done: cost/latency/disagreement dashboards, Langfuse/Phoenix ops hooks, and the Shadow->Verify->Skip rollout itself. Keeping open until the deployment thresholds are quantified and rollout begins.

**Comment by Exios66** [comment]
## 🚀 Unblocked — real checkpoint available for the eval harness

The #84 artifact shipped (2026-09-20): **https://huggingface.co/Lucius-Morningstar/mailroom-modernbert-classifier** (2-epoch ModernBERT-base, run `20260920-004937`, ~$3.91, under budget).

**Eval-relevant numbers (val, epoch 2):** doc_type window acc 0.857 · doc acc 0.799 · macro-F1 0.699 · **ECE 0.225** (calibration gate NOT met — the selective-risk report this issue's harness produces is exactly the evidence needed to set the real threshold).

**Ready to execute:**
```
uv run python training/eval_modernbert.py --checkpoint <bundle> --subset test --sample 50 --seed 42 --selective-risk
```
- Short-doc cohort agreement vs sorter, calibrated ECE, $/doc, p50/p95 latency → the Shadow phase inputs for #85 Phase 4
- Selective-risk report → supplies `bert_intake_min_confidence` for #90's gate (P3/F3)

**Sequence per current plan:** specialist audit of the 2-epoch run → config improvements (verified) → re-run → then this eval harness runs against the improved checkpoint. The harness itself can be validated now against the current checkpoint.

**Comment by Exios66** [comment]
**Eval + rollout review (2026-09-20) — cohort facts, metrics schema, rollout posture**

1. **Short-doc cohort (verified against mailroom-dataset @ 46a4d3c, ground_truth ⋈ default/doc_text on filename, 0 unmatched)**: ≤30,000 chars = **2,713/3,302 rows (82.2%)**; 270/323 test rows. ⚠️ **merger_agreement contributes 0/152 rows** → cohort is 4-class; add the long-doc contract/merger supplement (see #104). No messy/quality column exists — "clean" must replicate the clerk's `looks_messy` heuristic exactly (train/serve skew risk). ~7 near-dups via content_sha256 → track group_id.
2. **Evidence-first rollout posture**: the published artifact is `gate_met: false` (ECE 0.225, doc acc 0.799). **Shadow can ship now** (compute + log + never skip) against the PyTorch bundle; **Verify** needs the #104 eval-hardening (min-n ≥ 30 + Wilson CI, per-head ECE sidecar, single-vs-multi-window cohorts) + sorter agreement ≥ 95% steady; **Skip** is near-zero coverage until the subclass-head pass (#107) and calibration land — that is correct posture, not a delay.
3. **Metrics schema** (mirrors existing SCORE_CONFIGS discipline; registration tracked in #106 - KANBAN-061): per-doc `bert_pass` (BOOL), `bert_sorter_agreement` (BOOL), `bert_fail_soft` (BOOL), `bert_elapsed_ms` (NUMERIC), `fast_path_est_cost_usd` (NUMERIC, derived ms×W/kWh×$/kWh — **never** conflated with API `estimated_cost_usd`); reuse `class_correct`, `run_duration_seconds`, `total_tokens`. Langfuse: one trace per doc, `session_id=mr-bert-shadow-<stamp>`, tags `bert-intake`/`shadow`; dashboard panels for BERT-pass %, skip %, disagreement %, fail-soft %, extract accuracy (aggregate gates, not per-doc).
4. **Dashboards**: extend the synced set (`sync_dashboards.py`); the existing "Mailroom Quality/Performance" panels get BERT cohorts as separate panels so the local-compute cost never pollutes API spend trends.
5. **Existing eval-harness inventory**: mailroom-ml `training/eval_modernbert.py` (report-only; test-split 323; window+band ECE; selective-risk sweep; no langfuse/cost/agreement yet); llm-mailroom `run_pilot.py`/`run_quality_judges.py` pattern (session per run, deterministic trace ids, docs/reports/pilots). The shadow runner should be an llm-mailroom script following `run_vision_sweep.py`'s multi-config comparison shape.
6. **Dependencies**: #88/#90/#91/#101 → #104/#106 (eval evidence + scores) → Skip promotion. Per-domain promotion starts correspondence/short notices only AFTER #107 subclass trust.

**Comment by Exios66** [comment]
**Progress — doctrine slice landed; eval gates pending run-3 (in flight).**

- [x] **Agents README / intake doctrine + rollback documented** — `packages/llm-mailroom/docs/intake-bert-doctrine.md` (monorepo `29e216bd`): tier ladder + enforcement points, precedence rules (BERT > intake prior; advisory overrule; reviewer blindness), gate semantics (skip/verify/shadow + terminal `gate_outcome`), fail-open doctrine, rollback table (`MAILROOM_BERT_INTAKE=0` / mode demotion / <95% drift rule), promotion gates. AGENTS.md + docs/agents.md link it.
- [x] **Dashboard panels — tracked follow-up with owners named** — new #111: hazel-ui-software-master (Observatory BERT panels), hermes cron (drift watchdogs).
- [x] Eval suite path exists (mailroom-ml tests + entity-package harness); metrics schema = the #106 score set.
- [ ] **Shadow eval on the short-doc cohort** — BLOCKED on run-3 (spawned `fc-01M31FA8F0W5AC9E2QNJAJ7AS4`, 2 epochs, lr 2e-5, `--label-smoothing=0.05 --loss-lambda-dt=0.65 --weight-mode=sqrt-inverse --mlp-heads`, L4, est. $4.11); then ONNX re-export → parity → sidecar bundle → latency → eval harness run → sign-off numbers recorded HERE.
- [ ] Verify canary + skip allowlist sign-off — after the eval numbers land.

Doctrine text already carries the promotion gates so #92 closes the moment the numbers are recorded.

**Comment by Exios66** [comment]
**Closed with evidence — phase-4 eval + rollout prep complete (run-3 numbers in).**

**Eval & metrics (shadow data):**
- Short-doc cohort (≤30k chars = 82.2% of docs) eval on the held-out test: **doc_type acc 0.8947**, window ECE 0.0203, 504 windows
- Selective-risk: **threshold 0.93** (2% budget, min-n 30, Wilson bound, coverage 43.3%, budget_met) — the deployment gate (replaces BERT_INTAKE_MIN_CONFIDENCE once adopted)
- **Cost/latency** (measured, Apple M2 CPU fp32 — the local fast path): p50 75 ms @64 tok / 308 ms @256 / 1.63 s @1,024; ≈ **\/bin/zsh.0002–0.005/doc** at 40 W — 10–100× under the LLM sorter path; L4-GPU full-context numbers linked from #107
- **Disagreement/fail-soft/accuracy panels LIVE**: Observatory BERT panels shipped (#111, commit `9d3bcf7a` — pass/skip/disagreement/fail-soft/accuracy, shadow-mode, per-revision columns)

**Rollout state (ordered, documented):**
1. **Shadow** — active (BERT computed + sorter always runs; metrics only) ✓
2. **Verify** — gate + PASS/FAIL harness green (#90/#91 fixtures, live-mode flag verified in tests; P1–P7 pane) ✓
3. **Skip** — NOT enabled (by design): requires the per-class allowlist + cost/accuracy sign-off ⟶ **punted to the epic #85 needs_attention** with the numbers — a human decision, not a code gap.

**Rollback**: `MAILROOM_BERT_INTAKE=0` + mode flag demotion documented (`docs/intake-bert-doctrine.md`, `29e216bd`). Lanfuse/Phoenix ops hooks: observability spans + panels per the doctrine.

Remaining #92 child work (eval-harness hardening paths) closed under #104; artifact serving contract closed under #101.
